File: typescipt/src/compiler/cSharp/builtInFunctions/methodFunctionCall.ts


internal abstract class MethodFunctionCall : FunctionCall {
   public ExpressionFunction Function

   protected abstract string ClassName
   protected abstract string MethodName

   protected MethodFunctionCall(ExpressionFunction function) : base(function) {
     Function = function;
   }

   public override customMethodSyntax(context: ICompileFunctionContext): MemberDeclarationSyntax {
     return null;
   }

   public override callExpressionSyntax(context: ICompileFunctionContext): ExpressionSyntax {
     let arguments = GetArguments(context);

     return SyntaxFactory.InvocationExpression(
         SyntaxFactory.MemberAccessExpression(
           SyntaxKind.SimpleMemberAccessExpression,
           SyntaxFactory.IdentifierName(ClassName),
           SyntaxFactory.IdentifierName(MethodName)))
       .WithArgumentList(SyntaxFactory.ArgumentList(arguments));
   }

   protected abstract getArguments(context: ICompileFunctionContext): SeparatedSyntaxArray<ArgumentSyntax>;
}

File: dotnet/Lexy/Lexy.Compiler/Compiler/CSharp/BuiltInFunctions/MethodFunctionCall.cs
using Lexy.Compiler.Language.Expressions.Functions;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace Lexy.Compiler.Compiler.CSharp.BuiltInFunctions;

internal abstract class MethodFunctionCall<TExpressionFunction> : FunctionCall<TExpressionFunction>
    where TExpressionFunction : ExpressionFunction
{
    protected abstract string ClassName { get; }
    protected abstract string MethodName { get; }

    public override ExpressionSyntax CallExpressionSyntax(TExpressionFunction expressionFunction)
    {
        var arguments = GetArguments(expressionFunction);

        return SyntaxFactory.InvocationExpression(
                SyntaxFactory.MemberAccessExpression(
                    SyntaxKind.SimpleMemberAccessExpression,
                    SyntaxFactory.IdentifierName(ClassName),
                    SyntaxFactory.IdentifierName(MethodName)))
            .WithArgumentList(SyntaxFactory.ArgumentList(arguments));
    }

    protected abstract SeparatedSyntaxList<ArgumentSyntax> GetArguments(TExpressionFunction powerFunction);
}